File: hebe-python/uonet_request_signer_hebe/signer.py
from OpenSSL import crypto
import hashlib
import base64
import json
import re
import urllib
from datetime import datetime


def get_encoded_path(full_url):
    path = re.search(r'(api/mobile/.+)', full_url)
    if path is None:
        raise ValueError('The URL does not seem correct (does not match `(api/mobile/.+)` regex)')
    return urllib.parse.quote(path[1], safe='').lower()


def get_digest(body):
    if not body:
        return None
    
    m = hashlib.sha256()
    m.update(bytes(body, 'utf-8'))
    return base64.b64encode(m.digest()).decode('utf-8')


def get_headers_list(body, digest, canonical_url, timestamp):
    sign_data = [
        ['vCanonicalUrl', canonical_url],
        ['Digest', digest] if body else None,
        ['vDate', timestamp.strftime('%a, %d %b %Y %H:%M:%S GMT')]
    ]
    
    return (
        ' '.join(item[0] for item in sign_data if item),
        ''.join(item[1] for item in sign_data if item)
    )


def get_signature(data, private_key):
    data_str = json.dumps(data) if isinstance(data, dict) or isinstance(data, list) else str(data)
    pkcs8 = crypto.load_privatekey(crypto.FILETYPE_ASN1, base64.b64decode(private_key))
    signature = crypto.sign(pkcs8, bytes(data_str, 'utf-8'), 'RSA-SHA256')
    return base64.b64encode(signature).decode('utf-8')


def get_signature_values(fingerprint, private_key, body, full_url, timestamp):
    canonical_url = get_encoded_path(full_url)
    digest = get_digest(body)
    headers, values = get_headers_list(body, digest, canonical_url, timestamp)
    signature = get_signature(values, private_key)

    return (
        f'SHA-256={digest}' if digest else None,
        canonical_url,
        f'keyId="{fingerprint}",headers="{headers}",algorithm="sha256withrsa",signature=Base64(SHA256withRSA({signature}))'
    )

File: hebe-jvm/src/main/kotlin/signer.kt
package io.github.wulkanowy.signer.hebe

import com.migcomponents.migbase64.Base64
import java.net.URLEncoder
import java.security.KeyFactory
import java.security.Signature
import java.security.spec.PKCS8EncodedKeySpec
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.security.MessageDigest.getInstance as createSign

private fun getDigest(body: String?): String? {
    if (body == null) return null
    return Base64.encodeToString(createSign("SHA-256").digest(body.toByteArray()), false)
}

private fun getSignatureValue(values: String, privatePem: String): String {
    val keyFactory = KeyFactory.getInstance("RSA")
    val privateBytes = Base64.decode(privatePem)
    val privateSpec = PKCS8EncodedKeySpec(privateBytes)
    val privateKey = keyFactory.generatePrivate(privateSpec)

    val signature = Signature.getInstance("SHA256withRSA")
    signature.initSign(privateKey)
    signature.update(values.toByteArray())

    return Base64.encodeToString(signature.sign(), false)
}

private fun getEncodedPath(path: String): String {
    val url = ("(api/mobile/.+)".toRegex().find(path))
        ?: throw IllegalArgumentException("The URL does not seem correct (does not match `(api/mobile/.+)` regex)")

    return URLEncoder.encode(url.groupValues[0], "UTF-8").orEmpty().toLowerCase()
}

private fun getHeaders(digest: String?, canonicalUrl: String, timestamp: ZonedDateTime): MutableMap<String, String> {
    val headers = mutableMapOf<String, String>()
    headers["vCanonicalUrl"] = canonicalUrl
    if (digest != null) headers["Digest"] = digest
    headers["vDate"] = timestamp.format(DateTimeFormatter.RFC_1123_DATE_TIME)
    return headers
}

fun getSignatureHeaders(
    keyId: String,
    privatePem: String,
    body: String?,
    requestPath: String,
    timestamp: ZonedDateTime
): Map<String, String> {
    val canonicalUrl = getEncodedPath(requestPath)
    val digest = getDigest(body)
    val headers = getHeaders(digest, canonicalUrl, timestamp.withZoneSameInstant(ZoneId.of("GMT")))
    val headerNames = headers.keys.joinToString(" ")
    val headerValues = headers.values.joinToString("")
    val signatureValue = getSignatureValue(headerValues, privatePem)

    headers["Digest"] = "SHA-256=${digest}"
    headers["Signature"] = """keyId="$keyId",headers="$headerNames",algorithm="sha256withrsa",signature=Base64(SHA256withRSA($signatureValue))"""

    return headers
}
